wordpress-mobile/release-toolkit
# Add ios_generate_strings_file_from_code action & specs

[ Part of project ref paaHJt-2Ib-p2 ]

## What

Adds a new `ios_generate_strings_file_from_code` action and its specs & test data.

This is aimed to replace our `localize.py` script duplicated in all our app repos

## Draft… but Ready for first review

This PR is still a Draft because once this new action is validated, I will deprecate the `IosLocalizeProjectAction` (and the `self.localize_project` method from `ios_git_helper.rb` it calls) which currently end up calling the `localize.py` script in the client app repo. Before doing so I wanted to give us the opportunity to test this new action's basics.

### Reviewing the Unit Tests and covered cases

To test at a high level, verify that:
 - The Unit Tests Pass on CI, and rubocop is happy
 - The Unit Tests I wrote cover all the use cases you can think of, especially the use cases that we could have already encountered when running `localize.py` vs if we ran this action on the apps' codebase.
 - The documentation is clear enough.

### Testing on an app repo

 - Create a test branch, point your `fastlane/Pluginfile` to this branch of the toolkit
 - Run `Scripts/localize.py` from your app repo, commit the changes in the `Localizable.strings` generated
 - Add a lane to your `Fastfile` to call this new action, and make sure you not only pass `WordPress` but also the paths to `Pods/` as the list of files to scan – ideally only explicitly listing the a8c Pods you want to include (instead of everything in Pods)
 - Compare the generated `.strings` file from this action with the ones generated by `localize.py` and ensure they match (†)

ℹ️  PS: I haven't personally run those steps above (aka testing my action on an app repo) yet, which is also why I'm opening this Draft PR early so that you can help me test that part while I wrap things up (i.e. deprecate the old action, etc)

## Differences in behavior between `localize.py` and this new action

### Merge dance in `Localize.py` not replicated in the new action (by design)

See paaHJt-2KK-p2 (search `merge_with` in the doc): the `Localize.py` script does this merge dance to keep any English translation already present in the local `.strings` that would differ from the generated one; this new fastlane action doesn't, because we expect to get rid of the need for this by fixing the apps codebase to use `genstrings` as Apple intended.

(†) I know of 2 keys in WPiOS which currently don't have the same key vs value in the existing `.strings` file, and thus for which `localize.py` handles the case by merging the generated `.strings` file with the existing one (giving values from the latter the priority) – while my new action does not handle such merging logic and would override the translation from the existing `.strings` file.

This is by design, as I expect to fix this not via the tooling (aka by implementing the merge logic from `localize.py` in this new action) but instead by fixing the app codebase, using `NSLocalizedString("key", value: "value")` when we need key ≠ value, avoiding to put the burden and complexity on the tooling and instead using build-in mechanisms already provided by iOS and `genstrings`.

This is why you should expect a diff on those 2 strings when comparing the output of running `localize.py` vs this action on WPiOS, which should disappear once we fix the WPiOS codebase appropriately.

### Parsing Pods option now more explicit (by design)

 - In `Localize.py`, the main method has an `include_pods_and_frameworks` boolean parameter.
 - In the new action, instead of a boolean, I've opted to let the caller provide an arbitrary list of paths to parse. It's now up to the caller to provide e.g. `['WordPress', 'Pods/`]` as the list of files – or even better, `['WordPress', 'Pods/WordPressLogin', 'Pods/WordPressUI', …]` to be more explicit about which Pods we want to translate the strings of via the app translation pipeline

### Generation of Widgets' base keys for reference not replicated (by design)

See also paaHJt-2KK-p2 – especially notes about 4️⃣ in the details of each script's role

In addition to generating the `.strings` file in `WordPress/Resources` for GlotPress to import, the `Localize.py` script also [generates the `Base.lproj/Localizable.strings` files for 2 of the Widget targets](https://github.com/wordpress-mobile/WordPress-iOS/blob/develop/Scripts/localize.py#L156-L157).
Those 2 additional files were not imported by GlotPress, but were instead user later in the process, when we download the translations from GlotPress, to compare which keys from those translations we need to "redistribute" to the widgets, by comparing the keys in those 2 generated `.strings` files from `localize.py` with the keys pulled from GlotPress

This behaviour is not reproduced in this action, because we plan to address this once again by a change in the codebase – especially by making the widgets instead use the strings from the app target at runtime, using `NSLocalizedString(…, bundle: theHostAppBundle, …)` in the widgets' codebase instead of defaulting to looking the strings in their own bundle.

Especially see Step 2 of paaHJt-2J8-p2 for explanations on that plan.
I'll have to explicitly test that use case when I get to implementing Step 2 of the plan though, I'm not there yet.

### No file exclusion option in the new action (bug; will add later)

While writing this PRs description and list of differences I just realised that this action doesn't not allow for providing exceptions to the list of files to be parsed, while in `localize.py`, [we do a bunch of `grep -v` pipes after the `find` in oder to exclude some files](https://github.com/wordpress-mobile/WordPress-iOS/blob/develop/Scripts/localize.py#L137). This possibility is missing from the current implementation of my new action as of today. I'll see to add that feature in a subsequent commit on this PR tomorrow.
